src/duckdb.mjs

```diff
@@ -1,361 +1,234 @@
+        name: column_name,
+        type: getType(column_type),
+        databaseType: column_type
+      };
+    });
+  }
 
-      return this;
+  static async of(sources = {}, config = {}) {
+    const db = await createDuckDB();
+    if (config.query?.castTimestampToDate === undefined) {
+      config = {...config, query: {...config.query, castTimestampToDate: true}};
     }
+    await db.open(config);
+    await Promise.all(
+      Object.entries(sources).map(async ([name, source]) => {
+        if ("array" in source) { // array + options
+          const {array, ...options} = source;
+          await insertArray(db, name, array, options);
+        } else if ("file" in source) { // file + options
+          const {file, ...options} = source;
+          await insertFile(db, name, file, options);
+        } else if (source instanceof FileAttachment) { // bare file
+          await insertFile(db, name, source);
+        } else if (Array.isArray(source)) { // bare data
+          await insertArray(db, name, source);
+        } else {
+          throw new Error(`invalid source: ${source}`);
+        }
+      })
+    );
+    return new DuckDBClient(db);
+  }
+}
 
-    async insertArrowTable(name, table, options) {
-      const buffer = arrow.tableToIPC(table);
-      return this.insertArrowFromIPCStream(name, buffer, options);
+async function insertFile(database, name, file, options) {
+  const url = await file.url();
+  if (url.startsWith("blob:")) {
+    const buffer = await file.arrayBuffer();
+    await database.registerFileBuffer(file.name, new Uint8Array(buffer));
+  } else {
+    await database.registerFileURL(file.name, url);
+  }
+  const connection = await database.connect();
+  try {
+    switch (file.mimeType) {
+      case "text/csv":
+        await connection.insertCSVFromPath(file.name, {
+          name,
+          schema: "main",
+          ...options
+        });
+        break;
+      case "application/json":
+        await connection.insertJSONFromPath(file.name, {
+          name,
+          schema: "main",
+          ...options
+        });
+        break;
+      default:
+        if (file.name.endsWith(".parquet")) {
+          await connection.query(
+            `CREATE VIEW '${name}' AS SELECT * FROM parquet_scan('${file.name}')`
+          );
+        } else {
+          throw new Error(`unknown file type: ${file.mimeType}`);
+        }
     }
+  } finally {
+    await connection.close();
+  }
+}
 
-    async insertArrowFromIPCStream(name, buffer, options) {
-      const db = await this.db();
-      const conn = await db.connect();
-      await conn.insertArrowFromIPCStream(buffer, {
-        name,
-        schema: "main",
-        ...options
-      });
-      await conn.close();
+async function insertArray(database, name, array, options) {
+  const arrow = await loadArrow();
+  const table = arrow.tableFromJSON(array);
+  const buffer = arrow.tableToIPC(table);
+  const connection = await database.connect();
+  try {
```

Comment:
I think it would be good to pull out a method for inserting an arrow `Table`. This method would be used to insert data that is already in columnar form.

Reply:
Added in 71fbda3.